### DISCUSSION TITLE
duckdb python api parquet ROW_GROUP_SIZE issue

### DISCUSSION THREAD

**Discussion by pyjads** [duckdb/duckdb#8392] [Show and tell]
I am trying to execute a query for a file

```
conn.sql(
            f"COPY k TO \'{destination}\' (FORMAT \'PARQUET\', ROW_GROUP_SIZE 500, PARTITION_BY {str_partition}, FILENAME_PATTERN \'{file_pattern_appender}\', OVERWRITE_OR_IGNORE)"
        )
```

The row group size of the output file is not according to the query.

Below are the details of the partitioned output parquet file:

```
num of rows: 1349
num of row groups: 1
row groups:

row group 0:
--------------------------------------------------------------------------------
total byte size: 0
num of rows: 1349
```

**Comment by szarnyasg** [comment]
Hi @pyjads, thanks for reporting this. Is there any chance you can share the data (or a sufficient subset of it) for reproducibility?

**Comment by lnkuiper** [comment]
DuckDB processes data in batches of 2048 rows at a time, so we can't go lower than that reliably. For example, if you set a row group size of 3000, then DuckDB will most likely write 2 batches, so you will get row groups of 4096.

**Comment by pyjads** [comment]
@szarnyasg 
I cannot provide the data as it is sensitive data.

However, I ran below query on the following data, I have 2 questions related to it.

[sample data](https://github.com/datablist/sample-csv-files/blob/main/files/customers/customers-10000.csv). I replaced space between headers with _ (Last Name was replaced with Last_Name etc.)

1. Query:
```
conn.sql(
            f"COPY k TO \'{destination}\' (FORMAT \'PARQUET\', ROW_GROUP_SIZE {row_group_size}, FILENAME_PATTERN \'{file_pattern_appender}\', OVERWRITE_OR_IGNORE)"
        )
```
ROW_GROUP_SIZE was set to 3000 as mentioned by @lnkuiper 
Here no partition was performed. 
I get following parquet metadata
```

num of rows: 10000
created by: DuckDB

num of row groups: 4

row groups:

row group 0:
--------------------------------------------------------------------------------
total byte size: 0
num of rows: 4096

row group 1:
--------------------------------------------------------------------------------
total byte size: 0
num of rows: 4096

row group 2:
--------------------------------------------------------------------------------
total byte size: 0
num of rows: 912

num of columns: 12

row group 3:
--------------------------------------------------------------------------------
total byte size: 0
num of rows: 896
```

The number of rows in row group 2 and row group 3 are very less compared to 1 and 2. Also the number of rows in 1 and 2 still exceeds the row_group_size. Also, I did not understand why row group 3 was created instead of merging the rows of group 3 to group 2.

2. If I run below query now
```
row_group_size = 3000
partition_by= [Last_Name]
str_partition = '({})'.format(','.join(partition_by))

conn.sql(
            f"COPY k TO \'{destination}\' (FORMAT \'PARQUET\', ROW_GROUP_SIZE {row_group_size}, PARTITION_BY {str_partition}, FILENAME_PATTERN \'{file_pattern_appender}\', OVERWRITE_OR_IGNORE)"
        )
```


In the destination directory two parquet files were created in each partition. Below is the image
![image](https://github.com/duckdb/duckdb/assets/61117329/db24645b-4f1e-49b0-ae65-5d35ab1f025e)

Why duckdb is creating more than 1 file for a partition where number of rows in parquet file is less than 10? If I use pyarrow, it creates only one file.

Also if I use the filesystem to connect to azure blob storage to write these parquet files to azure storage container, it takes very long time compared to local filesystem. However, using pyarrow to upload to azure storage container is nearly 10x faster.

**Comment by pyjads** [comment]
Any update on this?

**Comment by l1t1** [comment]
>Why duckdb is creating more than 1 file for a partition 

maybe a worker wrote a file and many workers read data of a partition. 
`set threads=1;` to solve the question.
